### PR TITLE
Signing 2 - hc keygen

### PR DESCRIPTION
**Pull request by lucksus**
- [x] I have added a summary of my changes to the changelog

`hc keygen` asks for passhprase and writes encrypted key bundle file to `~/.holochain/keys`.

### CODE REVIEW

File: cli/src/cli/keygen.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            Some(PwHashConfig(
+                pwhash::OPSLIMIT_INTERACTIVE,
+                pwhash::MEMLIMIT_INTERACTIVE,
+                pwhash::ALG_ARGON2ID13,
+            )),
+        )
+        .unwrap();
+
+    let path = match directories::UserDirs::new() {
+        Some(user_dirs) => user_dirs.home_dir().join(".holochain").join("keys"),
```
Comment (Connoropolous) [inline cli/src/cli/keygen.rs]: Based on what I added into the tree about directories, I propose that we create a new crate, or shareable file, that keeps track of and documents all the “magic strings” representing paths on devices. I find that they get scattered in the code and the act as high risk points of failure for our system I think. If we could start in that direction in this Pr that would be amazing

Comment (sphinxc0re) [reply cli/src/cli/keygen.rs]: Also I'm in favor of creating these directories with the XDG_CONFIG convention in mind. The `directories` crate already does that for you, we just have to use it.

File: common/src/paths.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+/// Returns the path to the root config directory for all of Holochain.
+/// If we can get a user directory it will be an XDG compliant path
+/// like "/home/peter/.config/holochain".
+/// If it can't get a user directory it will default to "/etc/holochain".
+pub fn config_root() -> PathBuf {
+    directories::ProjectDirs::from(QUALIFIER, ORGANIZATION, APPLICATION)
+        .and_then(|dirs| Some(dirs.config_dir().to_owned()))
+        .or(Some(PathBuf::new().join("/etc").join(APPLICATION)))
+        .unwrap()
```
Comment (sphinxc0re) [inline common/src/paths.rs]: Better make this an `unwrap_or_else`:
```rust
unwrap_or_else(|| PathBuf::from("something"));
```
also the `/etc` path won't work on windows either

File: common/src/paths.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+pub const APPLICATION: &'static str = "holochain";
+pub const KEYS_DIRECTORY: &'static str = "keys";
+
+/// Returns the path to the root config directory for all of Holochain.
+/// If we can get a user directory it will be an XDG compliant path
+/// like "/home/peter/.config/holochain".
+/// If it can't get a user directory it will default to "/etc/holochain".
+pub fn config_root() -> PathBuf {
+    directories::ProjectDirs::from(QUALIFIER, ORGANIZATION, APPLICATION)
+        .and_then(|dirs| Some(dirs.config_dir().to_owned()))
```
Comment (sphinxc0re) [inline common/src/paths.rs]: This can be a `map`
```suggestion
        .map(|dirs| dirs.config_dir().to_owned())
```

File: common/src/paths.rs (hunk 2 above)
Comment (sphinxc0re) [reply common/src/paths.rs]: I think we should bubble up the error instead of choosing a dummy path

File: hc_dpki/src/lib.rs [hunk 4]
```diff
@@ -4,10 +4,11 @@ extern crate holochain_sodium;
 #[macro_use]
 extern crate arrayref;
 extern crate base64;
-extern crate rustc_serialize;
-
 extern crate bip39;
 extern crate boolinator;
+extern crate rustc_serialize;
```
Comment (sphinxc0re) [inline hc_dpki/src/lib.rs]: Why is this needed again?

**Comment by Connoropolous** [comment]
Also, this needs to be added to the readme of CLI, plus the mdbook.

File: hc_dpki/src/lib.rs (hunk 4 above)
Comment (lucksus) [reply hc_dpki/src/lib.rs]: I don't know. I've just moved this by two lines for alphanumeric sorting.

File: common/src/paths.rs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+pub const KEYS_DIRECTORY: &'static str = "keys";
+
+/// Returns the path to the root config directory for all of Holochain.
+/// If we can get a user directory it will be an XDG compliant path
+/// like "/home/peter/.config/holochain".
+/// If it can't get a user directory it will default to "/etc/holochain".
+pub fn config_root() -> PathBuf {
+    directories::ProjectDirs::from(QUALIFIER, ORGANIZATION, APPLICATION)
+        .map(|dirs| dirs.config_dir().to_owned())
+        .unwrap_or_else(|| PathBuf::new().join("/etc").join(APPLICATION))
```
Comment (sphinxc0re) [inline common/src/paths.rs]:
```suggestion
        .unwrap_or_else(|| PathBuf::from("/etc").join(APPLICATION))
```
Also I still think, this is a bad idea

Comment (lucksus) [reply common/src/paths.rs]: What is a bad idea?

Comment (sphinxc0re) [reply common/src/paths.rs]: Btw, proof, that these are the same:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=978badfa745d4fa9b744dbf8229b28e9

**Comment by Connoropolous** [comment]
Ok, this is great I think, AND it takes us for now to a point where there is a `~/.hc` `~/.holochain` and now additionally `~/.config/holochain` ... I guess this is with the intention to press forward on making this consistent, and the XDG compliant approach would be the one we'd consolidate to?

File: cli/src/main.rs [hunk 6]
```diff
@@ -142,6 +146,12 @@ enum Cli {
         #[structopt(long = "skip-package", short = "s", help = "Skip packaging DNA")]
         skip_build: bool,
     },
+    #[structopt(
+        name = "keygen",
+        alias = "k",
+        about = "Creates a new agent key pair, asks for a passphrase and writes an encrypted key bundle to ~/.holochain/keys"
```
Comment (Connoropolous) [inline cli/src/main.rs]:
```suggestion
        about = "Creates a new agent key pair, asks for a passphrase and writes an encrypted key bundle to ~/.config/holochain/keys"
```

**Comment by lucksus** [comment]
@sphinxc0re, ok to merge?

**Comment by Connoropolous** [comment]
@lucksus can you comment on https://github.com/holochain/holochain-rust/pull/974#issuecomment-462373731?

**Comment by lucksus** [comment]
@Connoropolous, yes, this only makes sense if we now switch all config path handling to `holochain_common::paths`. But please let us do that in a follow-up PR.
